src/consensus/tower.zig

```diff
@@ -0,0 +1,3177 @@
+            }
+            break :blk 0;
+        };
+
+        return ComputedBankState{
+            .voted_stakes = voted_stakes,
+            .total_stake = total_stake,
+            .fork_stake = fork_stake,
+            .lockout_intervals = lockout_intervals,
+            .my_latest_landed_vote = my_latest_landed_vote,
+        };
+    }
+
+    pub fn lastVotedSlotInBank(
+        allocator: std.mem.Allocator,
+        accounts_db: *AccountsDB,
+        vote_account_pubkey: *const Pubkey,
+    ) ?Slot {
+        const vote_account = accounts_db.getAccount(vote_account_pubkey) catch return null;
+        const vote_state = stateFromAccount(
+            allocator,
+            &vote_account,
+            vote_account_pubkey,
+        ) catch return null;
+        return vote_state.lastVotedSlot();
+    }
+
+    fn stateFromAccount(
+        allocator: std.mem.Allocator,
+        vote_account: *const Account,
+        vote_account_pubkey: *const Pubkey,
+    ) !VoteState {
+        const buf = try allocator.alloc(u8, vote_account.data.len());
+        // TODO Not sure if this is the way to get the data from the vote account. Review.
+        _ = vote_account.writeToBuf(vote_account_pubkey, buf);
+        const versioned_state = try sig.bincode.readFromSlice(
+            allocator,
+            VoteStateVersions,
+            buf,
+            .{},
+        );
+        return try versioned_state.convertToCurrent(allocator);
+    }
+
+    /// Checks if `maybe_descendant` is a descendant of `slot`.
+    ///
+    /// Returns none if `maybe_descendant` is not present in `ancestors`
+    fn isDescendantSlot(
+        maybe_descendant: Slot,
+        slot: Slot,
+        ancestors: *const AutoHashMapUnmanaged(Slot, SortedSet(Slot)),
+    ) ?bool {
+        return if (ancestors.get(maybe_descendant)) |candidate_slot_ancestors|
+            candidate_slot_ancestors.contains(slot)
+        else
+            null;
+    }
+
+    /// Returns `Some(gca)` where `gca` is the greatest (by slot number)
+    /// common ancestor of both `slot_a` and `slot_b`.
+    ///
+    /// Returns `null` if:
+    /// * `slot_a` is not in `ancestors`
+    /// * `slot_b` is not in `ancestors`
+    /// * There is no common ancestor of slot_a and slot_b in `ancestors`
+    fn greatestCommonAncestor(
+        ancestors: *const AutoHashMapUnmanaged(
+            Slot,
+            SortedSet(Slot),
+        ),
+        slot_a: Slot,
+        slot_b: Slot,
+    ) ?Slot {
+        var ancestors_a = ancestors.get(slot_a) orelse return null;
+        var ancestors_b = ancestors.get(slot_b) orelse return null;
+
+        var max_slot: ?Slot = null;
+
+        for (ancestors_a.items()) |slot| {
+            for (ancestors_b.items()) |other_slot| {
+                if (slot == other_slot and
+                    (max_slot == null or
+                    slot > max_slot.?))
+                {
+                    max_slot = slot;
+                }
+            }
+        }
```

Comment:
Possible improvement? (untested code, but you get the idea with some tweaks)
```suggestion
        const superset, const subset = if (ancestors_a.count() >= ancestors_b.count())
            .{ ancestors_a, ancestors_b }
        else
            .{ ancestors_b, ancestors_a };

        for (superset.items()) |slot| {
            if (!subset.contains(slot)) continue;
            max_slot = if (max_slot) |current| @max(current, slot) else slot;
        }

```

Reply:
[ab41792](https://github.com/Syndica/sig/pull/664/commits/ab41792a1626567db1b6d0779dca12e77b72f5b4)

Reply:
nice, good catch on checking if either set is empty